### ISSUE TITLE
CQL->ELM compiler leaves out crucial type information

### ISSUE THREAD

**Issue by ewoutkramer** [FirelyTeam/firely-cql-sdk#209]
The Java CQL->ELM compiler can annotate the ELM with the determined static types of the nodes in the ELM tree. However, there are quite a few occasions where these annotations are left out, sometimes at crucial points, e.g.

```
{
                 "name" : "ToString",
                 "libraryName" : "FHIRHelpers",
                 "type" : "FunctionRef",
                 "operand" : [ {
                    "path" : "system",
                    "type" : "Property",
                    "source" : {
                       "type" : "First",
                       "source" : {
                          "path" : "coding",
                          "type" : "Property",
                          "source" : {
                             "path" : "code",
                             "scope" : "$this",
                             "type" : "Property"
                          }
                       }
                    }
                 }
```

Note how *none* of the Property nodes in this expression have a type annotation, so the user of the ELM (in our case the ExpressionBuilder) has to redo the semantic analysis that the compiler has already done. Can we ask the Java team to fix this?

Most of the time, there is at least some information:

```
{
              "path" : "component",
              "type" : "Property",
              "source" : {
                 "locator" : "173:11-173:25",
                 "resultTypeName" : "{http://hl7.org/fhir}Observation",
                 "name" : "EncounterLastBP",
                 "type" : "QueryLetRef"
              }
           }
```
Here, we see that the result of the QueryLetRef has a type annotation, but then again, the property reference (`component`) has not.

If this cannot be fixed, we have to think about sharing some code between our to-be CQL->ELM compiler and the ExpressionBuilder to redo the analysis....

**Comment by EvanMachusak** [comment]
See [this code](https://github.com/FirelyTeam/firely-cql-sdk/blob/9d3cbd97369d6973b4e5ce2ddea6c62c3d074a32/Cql/Cql.Compiler/ExpressionBuilder.cs#L2439)

You are supposed to get the return type of these implicit conversions from the ModelInfo.  The return type of `ToString` is indicated [here](https://github.com/FirelyTeam/firely-cql-sdk/blob/9d3cbd97369d6973b4e5ce2ddea6c62c3d074a32/Cql/Cql.Model/Models/fhir-modelinfo-4.0.1.xml#L11809C1-L11809C120):

` <conversionInfo functionName="FHIRHelpers.ToString" fromType="FHIR.EnrollmentRequestStatus" toType="System.String"/>`

We should be asking them to fix this.  Note this isn't the only place you're going to see this.  CQL-to-ELM doesn't annotate types on some aliasref nodes either.  It's really annoying.

And also keep in mind that we have code in various places that supports older renditions of ELM that could still exist in the wild and we need to support.

**Comment by ewoutkramer** [comment]
It's not really FhirHelper that's the problem here. It's this bit:

```
 {
                    "path" : "system",
                    "type" : "Property",
                    "source" : {
                       "type" : "First",
                       (no type info here)
                   }
```

When generating the code for the property access for `system`, we try to derive the type for the element - which comes from the call to `First` - that is something we cannot do.  BUT, by the time we have reached this point, we have actually already generated the Linq.Expression for the call to `First`.  However, the `Property` function in the ExpressionBuilder insists on getting the type form the Element, not the generated Expression. I added a bit of code that trusts the type coming in from the Linq.Expression if we cannot figure it out from the Element. That fixes this *for now*, but should probably be improved.
